aiohttp/client.py

```diff
@@ -574,34 +576,40 @@ async def _request(
                             resp.release()
 
                         try:
-                            parsed_url = URL(
+                            parsed_redirect_url = URL(
                                 r_url, encoded=not self._requote_redirect_url
                             )
-
                         except ValueError as e:
-                            raise InvalidURL(r_url) from e
+                            raise InvalidRedirectUrl(r_url) from e
 
-                        scheme = parsed_url.scheme
+                        scheme = parsed_redirect_url.scheme
                         if scheme not in ("http", "https", ""):
                             resp.close()
-                            raise ValueError("Can redirect only to http or https")
+                            raise InvalidRedirectUrl(
+                                r_url, "Can redirect only to http or https"
+                            )
                         elif not scheme:
-                            parsed_url = url.join(parsed_url)
+                            parsed_redirect_url = url.join(parsed_redirect_url)
 
                         is_same_host_https_redirect = (
-                            url.host == parsed_url.host
-                            and parsed_url.scheme == "https"
+                            url.host == parsed_redirect_url.host
+                            and parsed_redirect_url.scheme == "https"
                             and url.scheme == "http"
                         )
 
+                        try:
+                            redirect_origin = parsed_redirect_url.origin()
+                        except ValueError as e:
+                            raise InvalidRedirectUrl(parsed_redirect_url) from e
```

Comment:
```suggestion
                            raise InvalidRedirectUrl(parsed_redirect_url)
```
from e is the default behaviour.

Reply:
Actually, they are subtly different, this is probably fine.

Reply:
Yup, traceback seems to be just a little bit more clear with `from`.